velox/common/base/AsyncSource.h

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include <folly/executors/QueuedImmediateExecutor.h>
+#include <folly/futures/Future.h>
+
+namespace facebook::velox {
+
+// A future-like object that prefabricates Items on an executor and
+// allows consumer threads to pick items as they are ready. If the
+// consumer needs the item before the executor started making it,
+// the consumer will make it instead. If multiple consumers request
+// the same item, exactly one gets it.
+template <typename Item>
+class AsyncSource {
+ public:
+  AsyncSource(std::function<std::unique_ptr<Item>()> make) : make_(make) {}
+
+  // Makes an item if it is not already made. To be called on a background
+  // executor.
+  void prepare() {
+    std::function<std::unique_ptr<Item>()> make = nullptr;
+    {
+      std::lock_guard<std::mutex> l(mutex_);
+      if (!make_) {
+        return;
+      }
+      making_ = true;
+      std::swap(make, make_);
+    }
+    item_ = make();
+    {
+      std::lock_guard<std::mutex> l(mutex_);
+      making_ = false;
+      if (promise_) {
+        promise_->setValue(true);
+        promise_ = nullptr;
+      }
+    }
+  }
+
+  // Returns the item to the first caller and nullptr to subsequent callers. If
+  // the item is preparing on the executor, waits for the item and otherwise
+  // makes it on the caller thread.
+  std::unique_ptr<Item> move() {
```

Comment:
another way to implement this method (in a truly async way), is making it return a Future instead of blocking internally. The client of this API could then decide whether to block or propagate that future. Internally you would have to hold a SharedPromise instead of a Promise